File: python/src/a1_factory_method.py
"""
The intent is to define an interface for creating an object, but let subclasses
decide which class to instantiate. Factory Method lets a class defer instantiation
to subclasses.

Key participants:
- Product: Defines the interface of objects the factory method creates
- ConcreteProduct: Implements the Product interface
- Creator: Declares the factory method that returns a Product object
- ConcreteCreator: Overrides the factory method to return a ConcreteProduct

Key benefits:
- Eliminates the need to bind application classes to specific classes
- Code only deals with the Product interface, not concrete products
- Easy to extend by creating new subclasses

Real-world example: A document processing application that can create different
types of documents (TextDocument, SpreadsheetDocument) based on the application type.
"""
from abc import ABC, abstractmethod

# ----- Product -----
class Document(ABC):
    """Abstract Product"""
    @abstractmethod
    def create(self) -> None:
        pass

    @abstractmethod
    def get_type(self) -> str:
        pass


# ----- Concrete Products -----
class TextDocument(Document):
    def create(self) -> None:
        print("[TextDocument] Creating new text document...")
        print("[TextDocument] Setting up text editor features...")
        print("[TextDocument] Document ready!")

    def get_type(self) -> str:
        return "Text Document"


class SpreadsheetDocument(Document):
    def create(self) -> None:
        print("[SpreadsheetDocument] Creating new spreadsheet...")
        print("[SpreadsheetDocument] Initializing grid and formulas...")
        print("[SpreadsheetDocument] Document ready!")

    def get_type(self) -> str:
        return "Spreadsheet Document"


# ----- Creator -----
class Application(ABC):
    """Abstract Creator"""

    def new_document(self):
        """Factory method - subclasses override this"""
        doc = self.create_document()
        doc.create()
        print(f"[Application] Created: {doc.get_type()}")
        return doc

    @abstractmethod
    def create_document(self) -> Document:
        pass


# ----- Concrete Creators -----
class TextEditorApp(Application):
    def create_document(self) -> Document:
        return TextDocument()


class SpreadsheetApp(Application):
    def create_document(self) -> Document:
        return SpreadsheetDocument()


# ----- Example usage -----
def factory_method_example(use_spreadsheet: bool = False) -> None:
    """
    Demonstrates the Factory Method pattern.

    Args:
        use_spreadsheet: If True, creates a SpreadsheetApp, otherwise TextEditorApp
    """
    if use_spreadsheet:
        app = SpreadsheetApp()
    else:
        app = TextEditorApp()

    print("\n--- Factory Method: Creating Document ---")
    app.new_document()


if __name__ == "__main__":
    factory_method_example(use_spreadsheet=False)
    print("\n" + "="*50 + "\n")
    factory_method_example(use_spreadsheet=True)

File: rust/src/a1_factory_method.rs
//! # Factory Method Pattern
//!
//! The Factory Method pattern defines an interface for creating an object,
//! but lets subclasses decide which class to instantiate.
//!
//! ## Key Participants
//! - **Product**: The interface for objects the factory method creates
//! - **ConcreteProduct**: Implements the Product interface
//! - **Creator**: Declares the factory method
//! - **ConcreteCreator**: Overrides the factory method to return ConcreteProduct
//!
//! ## Benefits
//! - Eliminates need to bind application classes to specific classes
//! - Code only deals with the Product interface
//! - Easy to extend for new products

// Product trait
#[allow(dead_code)]
pub trait Document {
    fn create(&self);
    fn get_type(&self) -> &str;
}

// Concrete Products
pub struct TextDocument {
    content: String,
}

#[allow(dead_code)]
impl TextDocument {
    pub fn new() -> Self {
        Self {
            content: String::new(),
        }
    }

    pub fn with_content(content: &str) -> Self {
        Self {
            content: content.to_string(),
        }
    }
}

impl Document for TextDocument {
    fn create(&self) {
        println!("TextDocument created");
        if !self.content.is_empty() {
            println!("  Content: {}", self.content);
        }
    }

    fn get_type(&self) -> &str {
        "TextDocument"
    }
}

pub struct SpreadsheetDocument {
    rows: u32,
    cols: u32,
}

impl SpreadsheetDocument {
    pub fn new() -> Self {
        Self {
            rows: 1000,
            cols: 26,
        }
    }
}

impl Document for SpreadsheetDocument {
    fn create(&self) {
        println!("SpreadsheetDocument created with {}x{} grid", self.rows, self.cols);
    }

    fn get_type(&self) -> &str {
        "SpreadsheetDocument"
    }
}

// Creator (Application)
pub trait Application {
    fn new_document(&self) -> Box<dyn Document>;
    fn create_document(&self) {
        let doc = self.new_document();
        doc.create();
    }
}

// Concrete Creators
pub struct TextEditorApplication;

impl Application for TextEditorApplication {
    fn new_document(&self) -> Box<dyn Document> {
        Box::new(TextDocument::with_content("Hello, World!"))
    }
}

pub struct SpreadsheetApplication;

impl Application for SpreadsheetApplication {
    fn new_document(&self) -> Box<dyn Document> {
        Box::new(SpreadsheetDocument::new())
    }
}

/// Example demonstrating Factory Method pattern
pub fn example() {
    println!("\n--- Factory Method Pattern Example ---\n");

    println!("1. Text Editor Application:");
    let text_app = TextEditorApplication;
    text_app.create_document();

    println!("\n2. Spreadsheet Application:");
    let sheet_app = SpreadsheetApplication;
    sheet_app.create_document();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_text_document_creation() {
        let doc = TextDocument::new();
        assert_eq!(doc.get_type(), "TextDocument");
    }

    #[test]
    fn test_spreadsheet_document_creation() {
        let doc = SpreadsheetDocument::new();
        assert_eq!(doc.get_type(), "SpreadsheetDocument");
        assert_eq!(doc.rows, 1000);
        assert_eq!(doc.cols, 26);
    }

    #[test]
    fn test_text_editor_creates_text_document() {
        let app = TextEditorApplication;
        let doc = app.new_document();
        assert_eq!(doc.get_type(), "TextDocument");
    }

    #[test]
    fn test_spreadsheet_app_creates_spreadsheet() {
        let app = SpreadsheetApplication;
        let doc = app.new_document();
        assert_eq!(doc.get_type(), "SpreadsheetDocument");
    }
}
